hypertrace-trace-enricher/hypertrace-trace-enricher-api/src/main/java/org/hypertrace/traceenricher/trace/util/StructuredTraceGraphBuilder.java

```diff
@@ -0,0 +1,47 @@
+package org.hypertrace.traceenricher.trace.util;
+
+import org.hypertrace.core.datamodel.StructuredTrace;
+import org.hypertrace.core.datamodel.shared.StructuredTraceGraph;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class StructuredTraceGraphBuilder {
+  private static final Logger LOG = LoggerFactory.getLogger(StructuredTraceGraphBuilder.class);
+
+  private static ThreadLocal<StructuredTraceGraph> cachedGraph = new ThreadLocal<>();
+  private static ThreadLocal<StructuredTrace> cachedTrace = new ThreadLocal<>();
+
+  public static StructuredTraceGraph buildGraph(StructuredTrace trace) {
+    if (cachedTrace.get() == null) {
+      LOG.info("Building structured trace graph. Reason: no cached trace");
+      StructuredTraceGraph graph = StructuredTraceGraph.createGraph(trace);
+      cachedTrace.set(StructuredTrace.newBuilder(trace).build());
+      cachedGraph.set(graph);
+      return graph;
+    }
+
+    if (!cachedTrace.get().getCustomerId().equals(trace.getCustomerId()) ||
+        !cachedTrace.get().getTraceId().equals(trace.getTraceId())) {
+      LOG.info("Building structured trace graph. Reason: cached trace and current trace doesn't not match");
+      StructuredTraceGraph graph = StructuredTraceGraph.createGraph(trace);
+      cachedTrace.set(StructuredTrace.newBuilder(trace).build());
+      cachedGraph.set(graph);
+      return graph;
+    }
+
+    // trace internally changed
+    if(cachedTrace.get().getEntityList().size()  != trace.getEntityList().size() ||
```

Comment:
When this scenario will occur ?

Reply:
Can we pull this out in a separate boolean function? say `hasTraceChanged`?

Reply:
What about the set of attributes and enriched attributes? If the attributes and enriched attributes change, the graph structure will stay the same, but will we get the latest set of attributes from the graph nodes?

Basically, Does it create a copy of the Event, or just keeps the reference to it?

Reply:
It is keeping reference. ApiTraceGraph is using `getRootEvents`, but it is not in the enrichment process. @skjindal93 Can you also re-check?

Reply:
For `hasTraceChanged` currently, remaining condition of functions - isTraceSame and doeTraceExists works together for this function. So, if we move out, we have to move those conditions too. So, thinking of keeping this as-is for now.